### DISCUSSION TITLE
How to track memory consumption of a application

### DISCUSSION THREAD

**Discussion by suab321321** [opcm/pcm#251] [Q&A] [answered]
Hello, I want to monitor real time memory consumption of a running c++ application (in linux).
I want to know if I can do that using command line utility `pcm-memory.x`?

I am able to do it using c++ api provided, but I want to know if I can monitor using command line utility?

Thanks

**Comment by suab321321** [comment]
@opcm

**Comment by opcm** [comment]
do you want to monitor memory consumption or memory bandwidth consumption? PCM can only monitor memory bandwidth. Please also Q7: https://github.com/opcm/pcm/blob/master/FAQ.md

**Comment by suab321321** [comment]
hey @opcm I want to monitor memory bandwidth consumption.

**Comment by suab321321** [comment]
@opcm what I have tried,
1. in one terminal I am running `./pcm-memory.x`
2. in another termminal running a very simple my c++ application, this application allocates memory via `malloc` in the interval of `5` seconds.
I should see some diff output in the output of `pcm-memory.x` right when my c++ application code is running? But I m not seeing any of diff output, this is what I see


![Screenshot from 2020-11-22 18-16-04](https://user-images.githubusercontent.com/52123334/99904109-e1f99180-2cee-11eb-95ed-a35792e6439f.png)

my sample c++ code
```
#include <unistd.h>
#include <iostream>

using namespace std;

int main(){
for(int i=0;i<1000;i++){
long* c = (long*)malloc(sizeof(long)*10);
cout<<"allocated"<<endl;
sleep(2);
}
}
```

**Comment by opcm** [comment]
the memory allocation alone does not generate memory bandwidth. You need to read/write the memory. Also the array/accessed data needs to be larger than the CPU cache size to trigger memory bandwidth. Try 200 MB or larger arrays.

**Comment by suab321321** [comment]
@opcm I tried to allocate 200 MB data (`1024*1024*200`) after this I tried to do memset and then free, even after doing I m not able to see any changes in `./pcm-memory.x`.

I would like to point out some more info, this is what my internal of `pcm-memory.cpp` looks
```
int memfrq=0, memch=0, memcap=0,maxf=0,numcore=0,insize=0;
   memcap = atoi(argv[1]);
   memfrq = atoi(argv[2]);
   memch = atoi(argv[3]);
   maxf = atoi(argv[4]);
   numcore = atoi(argv[5]);
   insize = atoi(argv[6]);
```
I am running the exec `pcm-memory` with parameters like this `./pcm-memory.x 1 2 3 4 5 6` because I dont understand the variables `memcap,memfreq` means so I just put random values

**Comment by ogbrugge** [comment]
Your compiler might optimize it away, and just one time may not be enough to see a change.

Von meinem iPad gesendet

**Comment by opcm** [comment]
okay. Your CPU (Kabylake in the screenshot) is not supported by pcm-memory. Please see https://github.com/opcm/pcm/blob/master/FAQ.md#q6

How did you enforce pcm-memory to run? Did you modify pcm source code!?

On your CPU you can monitor the read and write memory bandwidth in pcm.x (READ and WRITE metrics)

**Comment by suab321321** [comment]
@opcm I did not modify anything it just ran on my system though I need to turn off nmi_watchdog and load the msr module, but I think these are to be done in supported cpu model too. I will try to run the test of intel xeon and see if it works there. Anyways thanks for clearing my doubt

**Comment by opcm** [comment]
> @opcm I tried to allocate 200 MB data (`1024*1024*200`) after this I tried to do memset and then free, even after doing I m not able to see any changes in `./pcm-memory.x`.
> 
> 
> 
> I would like to point out some more info, this is what my internal of `pcm-memory.cpp` looks
> 
> ```
> 
> int memfrq=0, memch=0, memcap=0,maxf=0,numcore=0,insize=0;
> 
>    memcap = atoi(argv[1]);
> 
>    memfrq = atoi(argv[2]);
> 
>    memch = atoi(argv[3]);
> 
>    maxf = atoi(argv[4]);
> 
>    numcore = atoi(argv[5]);
> 
>    insize = atoi(argv[6]);
> 
> ```
> 
> I am running the exec `pcm-memory` with parameters like this `./pcm-memory.x 1 2 3 4 5 6` because I dont understand the variables `memcap,memfreq` means so I just put random values

this is definitely not pcm-memory code above. I am not sure where did you get it from...

BTW: I have a Kabylake CPU too. pcm-memory.x does not work on it (as expected).

**Comment by suab321321** [comment]
@opcm But this is what I have in my source code of pcm-memory.cpp
Can you give me the link of correct pcm-memory.cpp

**Comment by suab321321** [comment]
@opcm can we monitor a simple program using `pcm.x` or `pcm-memory.x`, as of now I think I m getting info about the whole system with `pcm-memory.x`, but what if I want to monitor a particular c++ application, it is possible to do it (without using apis, but via command line utility)

**Comment by opcm** [comment, marked as answer]
the answers are in the FAQ recommended above..
